dnsext-types/DNS/Types/Domain.hs

```diff
@@ -182,25 +183,38 @@ modifyMailbox f (Mailbox o l) = Mailbox (f o) (f l)
 
 ----------------------------------------------------------------
 
-putDomain :: Domain -> SPut
-putDomain (Domain o _) = putDomain' _period o
+-- | Canonical-form flag.
+-- For example, `Canonical` is used from DNSSEC extension.
+--
+-- ref. https://datatracker.ietf.org/doc/html/rfc4034#section-6.2 - Canonical RR Form
+data CanonicalFlag
+  = NoCanonical
+  | Canonical
+  deriving Show
+
+----------------------------------------------------------------
+
+putDomain :: CanonicalFlag -> Domain -> SPut
+putDomain NoCanonical (Domain o _) = putDomain' _period True  o
+putDomain Canonical   (Domain _ l) = putDomain' _period False l {- canonical form is lowercase and no name-compression. -}
 
-putMailbox :: Mailbox -> SPut
-putMailbox (Mailbox o _) = putDomain' _at o
+putMailbox :: CanonicalFlag -> Mailbox -> SPut
+putMailbox NoCanonical (Mailbox o _) = putDomain' _at True  o
+putMailbox Canonical   (Mailbox _ l) = putDomain' _at False l {- canonical form is lowercase and no name-compression. -}
 
-putDomain' :: Word8 -> RawDomain -> SPut
-putDomain' sep dom
+putDomain' :: Word8 -> Bool -> RawDomain -> SPut
```

Comment:
I'm not sure why `Bool` is used here. Why don't you use `CanonicalFlag` directly?

Reply:
https://github.com/kazu-yamamoto/dnsext/pull/13/commits/faf2131a34d207ce852fd169a190f2b4ca06a151